src/linearcombination.jl

```diff
@@ -39,106 +42,82 @@ julia> LinearMap(ones(Int, 3, 3)) + CS + I + rand(3, 3);
+        @boundscheck check_dim_mul(y, A, x)
+        return _lincombmul!(y, A, x, α, β)
     end
-    return y
 end
 
-else # 5-arg mul! is available for matrices
-
-# map types that have an allocation-free 5-arg mul! implementation
-const FreeMap = Union{MatrixMap,UniformScalingMap}
-
-function A_mul_B!(y::AbstractVector, A::LinearCombination{T,As}, x::AbstractVector) where {T, As<:Tuple{Vararg{FreeMap}}}
-    # no size checking, will be done by individual maps
-    A_mul_B!(y, A.maps[1], x)
-    for n in 2:length(A.maps)
-        mul!(y, A.maps[n], x, true, true)
-    end
-    return y
-end
-function A_mul_B!(y::AbstractVector, A::LinearCombination, x::AbstractVector)
-    # no size checking, will be done by individual maps
-    A_mul_B!(y, A.maps[1], x)
-    l = length(A.maps)
-    if l>1
-        z = similar(y)
-        for n in 2:l
-            An = A.maps[n]
-            if An isa FreeMap
-                mul!(y, An, x, true, true)
-            else
-                A_mul_B!(z, A.maps[n], x)
-                y .+= z
-            end
+@inline function _lincombmul!(y, A::LinearCombination{<:Any,FiveArg}, x, α::Number, β::Number)
+    if iszero(α) # trivial cases
+        iszero(β) && (fill!(y, zero(eltype(y))); return y)
+        isone(β) && return y
+        # β != 0, 1
+        rmul!(y, β)
+        return y
+    else
+        mul!(y, first(A.maps), x, α, β)
+        @inbounds for An in Base.tail(A.maps)
+            mul!(y, An, x, α, true)
         end
+        return y
     end
-    return y
 end
 
-function LinearAlgebra.mul!(y::AbstractVector, A::LinearCombination{T,As}, x::AbstractVector, α::Number=true, β::Number=false) where {T, As<:Tuple{Vararg{FreeMap}}}
-    length(y) == size(A, 1) || throw(DimensionMismatch("mul!"))
-    if isone(α)
-        iszero(β) && (A_mul_B!(y, A, x); return y)
-        !isone(β) && rmul!(y, β)
-    elseif iszero(α)
+@inline function _lincombmul!(y, A::LinearCombination{<:Any,ThreeArg}, x, α::Number, β::Number)
+    if iszero(α)
         iszero(β) && (fill!(y, zero(eltype(y))); return y)
         isone(β) && return y
         # β != 0, 1
         rmul!(y, β)
         return y
-    else # α != 0, 1
-        if iszero(β)
-            A_mul_B!(y, A, x)
-            rmul!(y, α)
-            return y
-        elseif !isone(β)
-            rmul!(y, β)
-        end # β-cases
-    end # α-cases
-
-    for An in A.maps
-        mul!(y, An, x, α, true)
+    else
+        mul!(y, first(A.maps), x, α, β)
+        l = length(A.maps)
+        if l>1
+            z = similar(y)
+            @inbounds for n in 2:l
+                An = A.maps[n]
+                muladd!(mulstyle(An), y, An, x, α, z)
+            end
+        end
+        return y
     end
-    return y
 end
 
-end # VERSION
+@inline muladd!(::Type{FiveArg}, y, A, x, α, _) = mul!(y, A, x, α, true)
+@inline function muladd!(::Type{ThreeArg}, y, A, x, α, z)
+    A_mul_B!(z, A, x)
+    y .+= isone(α) ? z : z .* α
```

Comment:
I am not sure dot fusion will work with this ternary operator in between, so I think an explicit `if` branch is more recommended.